pym/bob/webdav.py

```diff
@@ -82,34 +61,49 @@ def _getHeaders(self):
                 userPass.encode("utf-8")).decode("ascii")
         return headers
 
+    def _getURL(self, path):
+        # remove username and password from URI
+        url = self.__url.scheme + "://" + self.__url.hostname
+        if self.__url.port is not None:
+             url += ":" + str(self.__url.port)
+        url += "/" + path
+        return url
+
     def exists(self, path):
-        connection = self._getConnection()
-        connection.request("HEAD", path, headers=self._getHeaders())
-        response = connection.getresponse()
-        response.read()
-        if response.status == 200:
-            return True
-        elif response.status == 404:
-            return False
-        else:
-            raise HttpUploadError("HEAD {} {}".format(response.status, response.reason))
+        req = urllib.request.Request (self._getURL(path),
+                                      headers=self._getHeaders(), method="HEAD")
+        try:
+            with urllib.request.urlopen (req, context=self.__context) as response:
+                if response.status == 200:
+                    return True
+        except urllib.error.HTTPError as e:
+            e.close()
+            if e.status != 404:
+                raise HttpUploadError("HEAD {} {}".format(e.status, e.reason))
+        return False
 
     def download(self, path, offset=None, length=None):
-        connection = self._getConnection()
         headers = self._getHeaders()
         if offset is not None and length is not None:
             headers.update({'Range': 'bytes={}-{}'.format(offset, offset + length - 1)})
-        connection.request("GET", path, headers=headers)
-        response = connection.getresponse()
-        if response.status in [200, 206]:
-            return response
-        else:
-            response.read()
-            if response.status == 404:
+
+        req = urllib.request.Request (self._getURL(path),
+                                      headers=headers, method="GET")
+        try:
+            resp = urllib.request.urlopen (req, context=self.__context)
+            if resp.status in [200, 206]:
+                return resp
+            else:
+                raise HttpDownloadError("{} {}".format(resp.status,
```

Comment:
I think `resp.close()` must be done here. But thinking more about it, I guess the case cannot happen at all. The server replied with a 2xx code to a GET request. I cannot imagine any other response than 200 and 206. So I would remove this case here.